### ISSUE TITLE
Unclear rule of ED & provider count in pallet-balances

### ISSUE THREAD

**Issue by xlc** [paritytech/substrate#14425]
With #12951, the rule of how ED and provider is changed and this breaks some of our code. While we are updating our code to work with the new behaviour, we found the existing code are preventing us to do what we believe a valid use case that should work.

Our use case:

Acala EVM+ requires a storage deposit to deploy smart contracts. The fund is reserved on the contract account and refunded when the contract is removed. The initial storage deposit is relatively large and always higher than ED. As new bytes are written in the contract storages, additional storage deposit will be charged and reserved on the contract account. Those additional deposits can be low and less than ED. We were expecting the initial deposit provides provider ref count to allow the contract to be able to receive additional storage deposit that's below ED.

Compatibility solution (that's not working, but we expect it to work):

Reserve no longer provides provider ref count. That's fine, as we can manually increase the provider ref count in our EVM pallet before transfer the initial storage deposit and reduce it after the contract is removed.

https://github.com/AcalaNetwork/Acala/issues/2495

However, the current implementation does not allow such case.

https://github.com/paritytech/substrate/blob/569aae5341ea0c1d10426fa1ec13a36c0b64393b/frame/system/src/lib.rs#L1689

This check will prevent transfer of any amount less than ED, regardless of provider count.

Is this behaviour expected? A provider ref count cannot lift the requirement of minimal balance?

Can we customize this behaviour? How (other than fork the code)?

To my understanding, reserve/hold no longer provides ref count because they are slashable. However, many use cases of reserve/hold does not intent to slash the funds and it makes perfect sense in such case to allow those funds to provide a ref count.  The requirement of a single pallet (staking) shouldn't impact all the other pallets and other parachains.

And do we have documentations of the minimal balances rule (of pallet balances and pallet assets) and provider / consume count rule and how they work together?

**Comment by gavofyork** [comment]
> And do we have documentations of the minimal balances rule (of pallet balances and pallet assets) and provider / consume count rule and how they work together?

This is documented in `frame_system`. Consumers cannot be created on an account without there being at least one provider. Providers cannot be removed from an account until there are zero consumers. A non-zero provider count or a non-zero sufficient count keep an account in existence.

Balances pallet gives a provider ref to accounts which contain funds. It disallows balances below the existential deposit. If a balance falls below, it is wiped out and there is no provider reference.

> Can we customize this behaviour?

No.

**Comment by xlc** [comment]
We are explicitly adding a provider count in our evm pallet so am I correct that with an account with provider count of 1, it is valid to transfer balances less than ED to it? Because that's not the case currently. 

This line does not check provider at all https://github.com/paritytech/substrate/blob/569aae5341ea0c1d10426fa1ec13a36c0b64393b/frame/system/src/lib.rs#L1689

**Comment by athei** [comment]
I remember that this behavior was counterintuitive for me, too. I was expecting that a manual  increment of a provider would lift the ED requirement and pull it into existence. But you still need the ED and hence the provider of `pallet-balances`.

**Comment by xlc** [comment]
can I say this is a bug that should be fixed?

**Comment by gavofyork** [comment]
The current logic is well-defined and is as expected:

> Balances pallet gives a provider ref to accounts which contain funds. It disallows balances below the existential deposit. If a balance falls below, it is wiped out and there is no provider reference.

For Balances pallet to allow free balances below the ED when providers are non-zero it would need to act as a consumer to be economically safe. It is not currently possible to do this since Balances pallet would need to switch between being a Consumer and a Provider depending on whether the balance is less than ED or not. When reducing the balance to somewhere between zero and the ED (not inclusive) it would need to determine whether other providers exist and adapt its behaviour between wiping out the account (if none do) or increment consumers (if there are some).

It would be possible to alter the logic to do this and I'd not be against including it. However, any such PR would obviously need a comprehensive review and audit before merge, release and upgrade. I'd expect the lead time to be no less than 2 months. As such I would seriously consider alternatives such as a path similar to that which the `pallet-contracts` team took.

**Comment by xlc** [comment]
I see. We would use different approach in our code to not be blocked by this. 

I will leave this open as I still want to see this improved.

**Comment by xlc** [comment]
I think a better solution will be refactoring pallet-balances in such way to allow customization of the behaviour of providers & dusts.

For example, I would like to customize it in such a way that with provider, it is possible for account to old balances less than ED without increase a consumer. When the provider is dropped to zero, the dust amount will be dusted.

**Comment by gavofyork** [comment]
How exactly would it be dusted? There is no callback in `frame_system::dec_provider` on purpose due to the difficulty with weight benchmarking that such arbitrary logic brings.

**Comment by gavofyork** [comment]
> I see. We would use different approach in our code to not be blocked by this.
> 
> I will leave this open as I still want to see this improved.

I'm not sure what the issue is; perhaps you want to change the name and description since I believe I've made the ED rule quite clear(?)

**Comment by xlc** [comment]
We already have an on account killed hook which should be enough.

Will make another issue.
